File: ios/Plugin/PointSDKPlugin.swift
import Foundation
import Capacitor
import PointSDK

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(PointSDKPlugin)
public class PointSDKPlugin: CAPPlugin {
    
    var healthKit: HealthKitManager?
    var healthService: HealthDataService { Point.healthDataService }
    
    @objc
    public func setup(_ call: CAPPluginCall) {
        Point.verbose = call.getBool("verbose", false)
        
        Point.setup(
            clientId: call.getString("clientId")!,
            clientSecret: call.getString("clientSecret")!,
            environment: environmentsMapping(call.getString("environment"))
        )
        
        call.resolve()
    }
    
    @objc
    public func setUserToken(_ call: CAPPluginCall) {
        Task {
            guard !Task.isCancelled else { return }
            
            do {
                try await Point.setAccessToken(
                    accessToken: call.getString("userToken")!,
                    shouldSyncHistoricalData: call.getBool("shouldSyncData", true)
                )
                call.resolve()
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc
    public func setAccessToken(_ call: CAPPluginCall) {
        Task {
            guard !Task.isCancelled else { return }
            
            do {
                try await Point.setAccessToken(
                    accessToken: call.getString("accessToken")!,
                    shouldSyncHistoricalData: call.getBool("shouldSyncData", true)
                )
                call.resolve()
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc
    public func setRefreshToken(_ call: CAPPluginCall) {
        Task {
            guard !Task.isCancelled else { return }
            
            do {
                let userId = call.getString("userId")!.replacingOccurrences(of: "|", with: "%7C", options: .literal, range: nil)
                try await Point.setRefreshToken(
                    refreshToken: call.getString("refreshToken")!,
                    userId: userId,
                    shouldSyncHistoricalData: call.getBool("shouldSyncData", true)
                )
                call.resolve()
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc
    public func logout(_ call: CAPPluginCall) {
        Task {
            guard !Task.isCancelled else { return }
            
            do {
                try await Point.logout()
                call.resolve()
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }
}

File: android/src/main/java/co/areyouonpoint/pointsdk/capacitor/PointSDKPlugin.kt
package co.areyouonpoint.pointsdk.capacitor

import co.areyouonpoint.pointsdk.PointClient
import co.areyouonpoint.pointsdk.core.logger.LogLevel
import co.areyouonpoint.pointsdk.domain.PointEnvironment
import co.areyouonpoint.pointsdk.domain.exceptions.PointException
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "PointSDK")
class PointSDKPlugin : Plugin() {
    private var pointClient: PointClient? = null

    @PluginMethod
    fun setup(call: PluginCall) {
        val verbose = call.getBoolean("verbose") ?: false

        pointClient = PointClient.getInstance(
            context = context,
            clientId = call.getString("clientId")!!,
            clientSecret = call.getString("clientSecret")!!,
            apiEnvironment = environmentsMapping(call.getString("environment")),
            logLevel = if (verbose) LogLevel.DEBUG else LogLevel.NONE
        )

        call.resolve()
    }

    @PluginMethod
    fun setUserToken(call: PluginCall) {
        try {
            pointClient?.setUserToken(call.getString("userToken")!!)
            call.resolve()
        } catch (ex: PointException) {
            call.reject(ex.message)
        }
    }

    @PluginMethod
    fun logout(call: PluginCall) {
        try {
            pointClient?.logout()
            call.resolve()
        } catch (ex: PointException) {
            call.reject(ex.message)
        }
    }

    private fun environmentsMapping(env: String?): PointEnvironment {
        return when(env) {
            "development" -> PointEnvironment.DEVELOPMENT
            "staging" -> PointEnvironment.STAGING
            "production" -> PointEnvironment.PRODUCTION
            "preprod" -> PointEnvironment.PRE_PROD
            else -> PointEnvironment.DEVELOPMENT //DEVELOPMENT IS THE DEFAULT SAME AS iOS
        }
    }
}
